### PR TITLE
Add option to parse all numbers as strings

### PR DESCRIPTION
**Pull request by mabranon**
This library used to automatically parse long integers as strings if they were unsafe to parse (because we would lose precision).

The implementation caused some issues because only unsafe numbers would be returned as strings (leading to unexpected behavior), and we also saw cases where numbers that were too large were not returned as strings because they coincidentally parsed to the same number (leading to more unexpected behavior). 

Now, an option object can be passed to the parser specifying that ALL numbers should be parsed as strings, and if the option is not given and an unsafe number is encountered an error is throw. 

{{whendone}}
update https://github.com/roirevolution/engineering/issues/274
{{/whendone}}

### CODE REVIEW

File: test/primitives.js [hunk 1]
```diff
@@ -23,14 +23,40 @@ var expected = [
   [ [], [ -1 ] ],
   [ [ 0 ], -0.1 ],
   [ [], [ -0.1 ] ],
-  [ [ 0 ], 6.02e+23 ],
-  [ [], [ 6.02e+23 ] ],
+  [ [ 0 ], 6.02e+10 ],
+  [ [], [ 6.02e+10 ] ],
```
Comment (mabranon) [inline test/primitives.js]: The old number was too big to be safe as a number.

File: test/primitives.js [hunk 2]
```diff
@@ -23,14 +23,40 @@ var expected = [
...
+];
+
+var stringExpected = [
```
Comment (mabranon) [inline test/primitives.js]: Could we move this into the test so that we don't need extra context in the variable name and call just call it `expected`.

File: test/primitives.js [hunk 3]
```diff
@@ -23,14 +23,40 @@ var expected = [
...
+  [ [ 1 ], '1.1' ],
+  [ [ 2 ], '-1.1' ],
+  [ [ 3 ], '-1.0' ],
+  [ [], [ '1.0', '1.1', '-1.1', '-1.0' ] ],
+  [ [ 0 ], '-1' ],
+  [ [], [ '-1' ] ],
+  [ [ 0 ], '-0.1' ],
+  [ [], [ '-0.1' ] ],
+  [ [ 0 ], '6.02e23' ],
+  [ [], [ '6.02e23' ] ],
```
Comment (mabranon) [inline test/primitives.js]: Could we add tests for:
- Really big negative numbers
- Negative exponents (and maybe quickly verify that they can be parsed as numbers on the caller's side?)
- Really big numbers (+/-) that have values after a decimal point
- Exponents without a decimal place
- Numbers too big (+/-) to be represented with a floating point number

File: test/primitives.js [hunk 4]
```diff
@@ -23,14 +23,40 @@ var expected = [
...
   [ [ 0 ], '7161093205057351174' ],
   [ [], [ '7161093205057351174'] ]
 ];
 
 test('primitives', function (t) {
-  t.plan(25);
+  t.plan(23);
```
Comment (mabranon) [inline test/primitives.js]: Can we use `t.end()` instead?

File: test/primitives.js [hunk 5]
```diff
@@ -45,6 +71,56 @@ test('primitives', function (t) {
...
+    p.write('[7161093205057351174]')
+
+    t.fail('Incorrectly attempted to parse unsafe large number')
+  } catch (e) {
+    t.assert(e.message.includes(`unsafe to parse as a number`))
+    t.end()
+  }
+});
+
+test('will emit an error if it attempts to parse an integer that is too small', function (t) {
```
Comment (mabranon) [inline test/primitives.js]: This is testing a very large negative number, I wouldn't call that "small."

File: jsonparse.js [hunk 6]
```diff
@@ -267,11 +268,15 @@ proto.write = function (buffer) {
...
-            if ((this.string.match(/[0-9]+/) == this.string) && (result.toString() != this.string)) {
-              // Long string of digits which is an ID string and not valid and/or safe JavaScript integer Number
-              this.onToken(STRING, this.string);
+            if (this.parseNumbersAsStrings) {
+              this.onToken(STRING, this.string)
             } else {
-              this.onToken(NUMBER, result);
+              if (result > Number.MAX_SAFE_INTEGER || result < Number.MIN_SAFE_INTEGER || !isFinite(result)) {
+                // Long string of digits which is an ID string and not valid and/or safe JavaScript integer Number
+                this.onError(new Error( `${this.string} is unsafe to parse as a number because it is either too large, too small, or not finite. Please pass the {parseNumbersAsStrings: true} option`));
```
Comment (mabranon) [inline jsonparse.js]: - too large to be supported as a finite number in javascript
- too large negative rather than too small

File: jsonparse.js [hunk 7]
```diff
@@ -51,9 +51,10 @@ var TAB =             "\t".charCodeAt(0);
 
 var STRING_BUFFER_SIZE = 64 * 1024;
 
-function Parser() {
+function Parser(options = {}) {
```
Comment (mabranon) [inline jsonparse.js]: We should make sure we're using a node version that supports defaults.

Comment (mabranon) [reply jsonparse.js]: Could we include the options in the default values? Something like this:
```suggestion
function Parser({parseNumbersAsString = false} = {}) {
```

File: jsonparse.js [hunk 8]
```diff
@@ -267,11 +268,15 @@ proto.write = function (buffer) {
               return this.charError(buffer, i);
             }
 
-            if ((this.string.match(/[0-9]+/) == this.string) && (result.toString() != this.string)) {
-              // Long string of digits which is an ID string and not valid and/or safe JavaScript integer Number
-              this.onToken(STRING, this.string);
+            if (this.parseNumbersAsStrings) {
```
Comment (mabranon) [inline jsonparse.js:271]: Can we do this before trying to parse the string as a number?

Comment (mabranon) [reply jsonparse.js:271]: Is there a good way to make sure this is a valid number, without converting it to a javascript number? Maybe using a regex or something?

File: jsonparse.js [hunk 9]
```diff
@@ -267,11 +268,15 @@ proto.write = function (buffer) {
...
+              this.onToken(STRING, this.string)
             } else {
-              this.onToken(NUMBER, result);
+              if (result > Number.MAX_SAFE_INTEGER || result < Number.MIN_SAFE_INTEGER || !isFinite(result)) {
```
Comment (mabranon) [inline jsonparse.js]: We should also throw errors for floats that are too precise.

File: jsonparse.js [hunk 10]
```diff
@@ -261,17 +266,19 @@ proto.write = function (buffer) {
             break;
           default:
             this.tState = START;
-            var result = Number(this.string);
-
-            if (isNaN(result)){
+            if (isNaN(this.string)){
```
Comment (mabranon) [inline jsonparse.js:267]: [isNaN will first coerce non-numeric values into numbers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isNaN#Confusing_special-case_behavior). This means that this check will pass if we get and empty string or a boolean.

This behavior should match what was already happening. Plus, any weird behavior here *should* be caught by `precisionWasLost` since we do a direct comparison to the value in the JSON (e.g. if the JSON value is `false`, `Number()` will convert that to `0`, and `precisionWasLost` should return true).

File: package.json [hunk 11]
```diff
@@ -1,22 +1,27 @@
...
+    "tape": "^4.9.0",
     "tap": "~0.3.3"
   },
   "scripts": {
     "test": "tap test/*.js"
   },
   "bugs": "http://github.com/creationix/jsonparse/issues",
-  "engines": ["node >= 0.2.0"],
+  "engines": [
+    "node >= 0.2.0"
```
Comment (amcheshire) [inline package.json]: Did we look into this? I believe there was concern that optional parameters weren't supported in this really old node version?

File: test/invalidNumber.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+})
+
+test('determines that precision is not lost for a safely precise negative integer', t => {
+  const jsonString = '-123908471290'
+  t.assert(!precisionWasLost(jsonString, Number(jsonString)))
+
+  t.end()
+})
+
+test('determines that precision is lost for an unsafely precise value in decimal notation', t => {
```
Comment (amcheshire) [inline test/invalidNumber.js]: For completeness, can we add a test that verifies it `'determines that precision is not lost for a safely precise value in decimal notation'`?

File: jsonparse.js [hunk 13]
```diff
@@ -408,6 +415,34 @@ proto.onToken = function (token, value) {
...
+ *
+ * The simplified significand only contains the digits and we remove any trailing zeros or decimal
+ * information.
+ * @param {string} numberString
+ * @return {string}
+ */
+function getDigitsSpecifyingPrecision(numberString) {
+  return numberString
+    .replace(/\./, '') // drops the dot if it exists
+    .replace(/^-?0?/, '') // drops leading negative sign and and leading 0's if present
```
Comment (amcheshire) [inline jsonparse.js]: Does this handle multiple leading zeroes? Example: `'00123'`.

Comment (amcheshire) [inline jsonparse.js]: Should we drop leading zeroes _before_ dropping the decimal point?

For example, `'.0012'` would currently become `'12'`, but I think we want it to be `'0012'` because leading zeroes _after_ the decimal point matter, right?

File: test/invalidNumber.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+})
+
+test('determines that precision is lost for a number that is too large (parses to infinity)', function (t) {
+  const jsonString = '5.98e10000'
+  t.assert(precisionWasLost(jsonString, Number(jsonString)))
+
+  t.end()
+})
+
+
```
Comment (amcheshire) [inline test/invalidNumber.js]: In general, how do you feel about using edge tests when possible? For example, have a test that verifies `Number.MAX_SAFE_INTEGER` is okay, but the next integer higher is not.

File: test/invalidNumber.js [hunk 15]
```diff
@@ -0,0 +1,76 @@
+var JsonParse = require('../jsonparse');
+var test = require('tape');
+
+const {precisionWasLost} = JsonParse.testing
+
+test('determines that precision is lost for an unsafely precise positive integer', t => {
```
Comment (mabranon) [inline test/invalidNumber.js]: Can we refactor these tests so that there is a const at the top that we loop over, either:

One const with the number and the expected output of `precision_was_lost`
two consts with the numbers that return true and the numbers that return false

File: test/invalidNumber.js [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+})
+
+test('determines that precision is lost for an unsafely precise negative integer', t => {
+  const jsonString = '-123908471290347091237490182734091237'
+  t.assert(precisionWasLost(jsonString, Number(jsonString)))
+
+  t.end()
+})
+
+test('determines that precision is not lost for a safely precise negative integer', t => {
```
Comment (mabranon) [inline test/invalidNumber.js]: Can we make it so we only have 1 negative test?

Also, if you happen to find one, can you add a test about a negative number that looses precision while the positive number doesn't?

File: test/invalidNumber.js [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+  t.end()
+})
+
+test('determines that precision is lost for a number that is too large (parses to infinity)', function (t) {
+  const jsonString = '5.98e10000'
+  t.assert(precisionWasLost(jsonString, Number(jsonString)))
+
+  t.end()
+})
```
Comment (mabranon) [inline test/invalidNumber.js]: Can we add a test for 
 - a large negative (-infinity)
 - and for a number that is too small (1/infinity)? 
 - A number with a lot of leading zeros
 - A number that is mostly zeros with just 2 ones (that looses precision)
 - A test for what happens with Nan / Null / Undefined / empty string

File: test/invalidNumber.js (hunk 14 above)
Comment (mabranon) [reply test/invalidNumber.js]: Depending on a lot of external considerations like the platform, MAX_SAFE_INTEGER + 1 could actually be parsed precisely.

File: test/primitives.js [hunk 18]
```diff
@@ -1,36 +1,32 @@
...
+    [ [ 1 ], 1.1 ],
+    [ [ 2 ], -1.1 ],
+    [ [ 3 ], -1 ],
+    [ [], [ 1, 1.1, -1.1, -1 ] ],
+    [ [ 0 ], -1 ],
+    [ [], [ -1 ] ],
+    [ [ 0 ], -0.1 ],
+    [ [], [ -0.1 ] ],
+    [ [ 0 ], 6.02e+10 ],
+    [ [], [ 6.02e+10 ] ],
```
Comment (mabranon) [inline test/primitives.js:28]: Let's add one of these with e+23 or higher?

File: test/primitives.js [hunk 19]
```diff
@@ -53,5 +116,14 @@ test('primitives', function (t) {
...
+  p.write('[5.02e-500]');
+  p.write('[-5.02e-500]');
   p.write('[7161093205057351174]');
+  p.write('[7161093205057351174.2189731287638]');
+  p.write('[-2098379823648235784297340123721386]')
+  p.write('[-2098379823648235784297340123721386.12983789123]')
+  p.write('[5e30]')
+
+  t.end()
 });
```
Comment (mabranon) [inline test/primitives.js:145]: Let's add a test to show that the parser throws.

File: package.json [hunk 20]
```diff
@@ -1,22 +1,27 @@
 {
   "name": "jsonparse",
   "description": "This is a pure-js JSON streaming parser for node.js",
-  "tags": ["json", "stream"],
+  "tags": [
```
Comment (mabranon) [inline package.json]: Let's reduce diffs

File: jsonparse.js [hunk 21]
```diff
@@ -408,6 +415,34 @@ proto.onToken = function (token, value) {
...
+    .replace(/[eE].*$/, '') // drops all exponent information
+    .replace(/0*$/, '') // drops trailing zeros
+}
+
 Parser.C = C;
 
 module.exports = Parser;
+module.exports.testing = testing
```
Comment (mabranon) [inline jsonparse.js]: Can this file import precisionWasLost from another file so that we can export this differently?

File: jsonparse.js (hunk 10 above)
Comment (mabranon) [reply jsonparse.js:267]: I would prefer to reduce the diff here, up to you.

File: test/primitives.js [hunk 22]
```diff
@@ -53,5 +116,14 @@ test('primitives', function (t) {
...
+  p.write('[-6.02e23]');
+  p.write('[5.02e-500]');
+  p.write('[-5.02e-500]');
   p.write('[7161093205057351174]');
+  p.write('[7161093205057351174.2189731287638]');
+  p.write('[-2098379823648235784297340123721386]')
+  p.write('[-2098379823648235784297340123721386.12983789123]')
+  p.write('[5e30]')
+
+  t.end()
```
Comment (mabranon) [inline test/primitives.js:144]: Can we ensure that there is a parse error for things like this:

```
{ key:   ,key2: value2}
```

Comment (mabranon) [reply test/primitives.js:144]: and this:

```
{key: 123hi, key2: value2}
```

File: jsonparse.js [hunk 23]
```diff
@@ -261,17 +266,19 @@ proto.write = function (buffer) {
...
-            if ((this.string.match(/[0-9]+/) == this.string) && (result.toString() != this.string)) {
-              // Long string of digits which is an ID string and not valid and/or safe JavaScript integer Number
-              this.onToken(STRING, this.string);
+            if (this.parseNumbersAsStrings) {
+              this.onToken(STRING, this.string)
             } else {
-              this.onToken(NUMBER, result);
+              const result = Number(this.string)
+              if (precisionWasLost(this.string, result)) {
+                this.onError(new Error( `${this.string} is unsafe to parse as a number because it is either too large (positive or negative), has too many significant figures, or is not finite when parsed. Please pass in the {parseNumbersAsStrings: true} option`))
```
Comment (mabranon) [inline jsonparse.js]:
```suggestion
                this.onError(new Error(`${this.string} is unsafe to parse as a number because it is either too large (positive or negative), has too many significant figures, or is not finite when parsed. Please pass in the {parseNumbersAsStrings: true} option`))
```

File: jsonparse.js [hunk 24]
```diff
@@ -408,6 +415,34 @@ proto.onToken = function (token, value) {
...
+/**
+ * Takes in a string represention of a number and returns a simplified version of the significand
+ * as a string.
+ *
+ * The simplified significand only contains the digits and we remove any trailing zeros or decimal
+ * information.
+ * @param {string} numberString
+ * @return {string}
+ */
+function getDigitsSpecifyingPrecision(numberString) {
```
Comment (mabranon) [inline jsonparse.js]:
```suggestion
function getSignificand(numberString) {
```

Comment (mabranon) [reply jsonparse.js]: Can we leave a link to the wikipedia page

Comment (mabranon) [reply jsonparse.js]: Can we clarify the assumed shape of this input? For example, we are assuming that if the number has any `., it has at most 1. 

It might just be "a number following the JSON specification"

File: test/invalidNumber.js (hunk 17 above)
Comment (mabranon) [reply test/invalidNumber.js]: Also, `+, -,` and empty string as signs for the exponent.

Comment (mabranon) [reply test/invalidNumber.js]: - Also, where the number after the `e` is actually too large for node to store (more than 64 bits).
 - Also, where the number after the `e` takes at least 20 bits to store.

Comment (mabranon) [reply test/invalidNumber.js]: - Also where the exponent starts with a 0?

File: jsonparse.js (hunk 13 above)
Comment (mabranon) [reply jsonparse.js]: I believe it will; however, we decided that we were only going to support what was supported by the JSON specification, so we will not explicitly support that case anyway

Comment (Tyler-Murphy) [reply jsonparse.js]: I think we want it to become `12` in that case. Leading zeroes are arbitrary, and there can be any number of them depending what power is in the exponent. For example, 12 is the same as 0.012e3.

File: test/primitives.js (hunk 1 above)
Comment (mabranon) [reply test/primitives.js]: This is no longer the case

File: precisionHelper.js [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+ * @param {string} jsonNumber
+ * @param {number} javascriptNumber
+ * @return {boolean}
+ */
+function precisionWasLost (jsonNumber, javascriptNumber) {
+  return getSignificand(jsonNumber) !== getSignificand(JSON.stringify(javascriptNumber))
+}
+
+/**
+ * Takes in a string represention of a number and returns a simplified version of the significand
```
Comment (Tyler-Murphy) [inline precisionHelper.js]: How about this?

`Takes in a JSON represention of a number and returns its significand`

I think we should drop the next line of the comment, and have just a link to wikipedia, where it's explained more clearly.

File: test/precisionWasLost.js [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+    expected: true,
+    description: 'precision is lost for a number that is too small'
+  },
+  {
+    jsonNumber: '100000000000000000000000010000',
+    expected: true,
+    description: 'precision is lost for a number that mostly zeros with just 2 ones'
+  },
+  {
+    jsonNumber: `5.2e${Number.MAX_VALUE}123`,
```
Comment (Tyler-Murphy) [inline test/precisionWasLost.js]: I don't think this is going to do what we want:

```js
> Number.MAX_VALUE.toString()
'1.7976931348623157e+308'
```

File: test/precisionWasLost.js [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+  {
+    jsonNumber: `5.2e09`,
+    expected: false,
+    description: 'precision is not lost for a safely precise number where the exponent starts with a leading 0'
+  },
+]
+
+test('precisionWasLost accurately detects loss of precision when a number represented as a string is parsed to a javascript number', t => {
+  valuesToTest.forEach(value => {
+    t.assert(precisionWasLost(value.jsonNumber, Number(value.jsonNumber)) === value.expected, value.description)
```
Comment (Tyler-Murphy) [inline test/precisionWasLost.js]: I think `tape` has an equality matcher:

```suggestion
    t.equal(precisionWasLost(value.jsonNumber, Number(value.jsonNumber)), value.expected, value.description)
```

**Review by Tyler-Murphy — DISMISSED** [review]
